common/tasks/sequential_scheduler.go

```diff
@@ -0,0 +1,328 @@
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tasks
+
+import (
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"go.temporal.io/server/common"
+	"go.temporal.io/server/common/backoff"
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/common/dynamicconfig"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+)
+
+var _ Scheduler[Task] = (*SequentialScheduler[Task])(nil)
+
+type (
+	SequentialSchedulerOptions struct {
+		QueueSize   int
+		WorkerCount dynamicconfig.IntPropertyFn
+	}
+
+	SequentialScheduler[T Task] struct {
+		status           int32
+		shutdownChan     chan struct{}
+		shutdownWG       sync.WaitGroup
+		workerShutdownCh []chan struct{}
+
+		options      *SequentialSchedulerOptions
+		queues       collection.ConcurrentTxMap
+		queueFactory SequentialTaskQueueFactory[T]
+		queueChan    chan SequentialTaskQueue[T]
+
+		logger log.Logger
+	}
+)
+
+func NewSequentialScheduler[T Task](
+	options *SequentialSchedulerOptions,
+	taskQueueHashFn collection.HashFunc,
+	taskQueueFactory SequentialTaskQueueFactory[T],
+	logger log.Logger,
+) *SequentialScheduler[T] {
+	return &SequentialScheduler[T]{
+		status:       common.DaemonStatusInitialized,
+		shutdownChan: make(chan struct{}),
+		options:      options,
+
+		logger: logger,
+
+		queueFactory: taskQueueFactory,
+		queueChan:    make(chan SequentialTaskQueue[T], options.QueueSize),
+		queues:       collection.NewShardedConcurrentTxMap(1024, taskQueueHashFn),
+	}
+}
+
+func (s *SequentialScheduler[T]) Start() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusInitialized,
+		common.DaemonStatusStarted,
+	) {
+		return
+	}
+
+	s.startWorkers(s.options.WorkerCount())
+
+	s.shutdownWG.Add(1)
+	go s.workerMonitor()
+
+	s.logger.Info("sequential scheduler started")
+}
+
+func (s *SequentialScheduler[T]) Stop() {
+	if !atomic.CompareAndSwapInt32(
+		&s.status,
+		common.DaemonStatusStarted,
+		common.DaemonStatusStopped,
+	) {
+		return
+	}
+
+	close(s.shutdownChan)
+	// must be called after the close of the shutdownChan
+	s.drainTasks()
+
+	go func() {
+		if success := common.AwaitWaitGroup(&s.shutdownWG, time.Minute); !success {
+			s.logger.Warn("sequential scheduler timed out waiting for workers")
+		}
+	}()
+	s.logger.Info("sequential scheduler stopped")
+}
+
+func (s *SequentialScheduler[T]) Submit(task T) {
+	queue := s.queueFactory(task)
+	queue.Add(task)
```

Comment:
can queueFactory function also do `queue.Add(task)` internally?

Reply:
Yes. But I prefer to have it as the queue factory is to generate the queue.